Add metrics-server

olga-mir opened the pull request:
```release-note
Add metrics-server
```

jim-minter commented on pkg/sync/v13/data/APIService.apiregistration.k8s.io/v1beta1.metrics.k8s.io.yaml:
```diff
@@ -0,0 +1,13 @@
+spec:
+  group: metrics.k8s.io
+  groupPriorityMinimum: 100
+  insecureSkipTLSVerify: true
```
I'm reluctant to merge with this set.

olga-mir replied:
thanks, updated

jim-minter commented on pluginconfig/pluginconfig-311.yaml:
```diff
@@ -56,6 +56,7 @@ versions:
       sync: quay.io/openshift-on-azure/azure:v7.0
       tlsProxy: quay.io/openshift-on-azure/azure:v7.0
       logAnalyticsAgent: mcr.microsoft.com/azuremonitor/containerinsights/ciprod:ciprod07092019
+      metricsServer: registry.access.redhat.com/openshift3/ose-metrics-server:v3.11.153-2
```
don't update older versions - only the most recent

asalkeld replied:
you have to otherwise config validation fails.

jim-minter replied:
then fix the validation ;-) -- see how it was done in Leszek's PR

jim-minter commented:
please add an e2e that proves that HPA works after this PR.

olga-mir commented on pkg/config/v13/generate_test.go:
```diff
@@ -90,6 +90,7 @@ func testRequiredFields(cs *api.OpenShiftManagedCluster, t *testing.T) {
 	assert(c.Images.GenevaLogging != "", "azure logging image")
 	assert(c.Images.GenevaTDAgent != "", "azure TDAgent image")
 	assert(c.Images.MetricsBridge != "", "metrics-bridge image")
+	assert(c.Images.LogAnalyticsAgent != "", "log analytics agent image")
```
(reminder to self  - remove this from metrics-server pr)

jim-minter commented:
@olga-mir have squashed commits and pushed some test fixes to your branch to help you out.  This is looking good.  E2E next...

jim-minter commented on pkg/sync/v14/data/ClusterRole.rbac.authorization.k8s.io/system-metrics-server.yaml:
```diff
@@ -0,0 +1,24 @@
+- apiGroups:
+  - extensions
+  resources:
+  - deployments
```
this seems odd - do we know why it's here?

olga-mir replied:
you are right this is not heeded, I will remove it. 
It got here from OpenShift old fork manifest: https://github.com/openshift/kubernetes-metrics-server/blob/master/deploy/resource-reader.yaml#L16
It was removed later in the upstream: https://github.com/kubernetes-sigs/metrics-server/pull/195

jim-minter commented on pkg/sync/v14/data/RoleBinding.rbac.authorization.k8s.io/kube-system/metrics-server-auth-reader.yaml:
```diff
@@ -0,0 +1,13 @@
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: Role
+  name: extension-apiserver-authentication-reader
```
as a one-off please can you manually check the apiserver audit logs to prove that the metrics server is pulling the extension-apiserver-authentication configmap?

olga-mir replied:
I can see in the metrics-server restarts on startup that it attempts to read it, but it was not yet available:

```
W1127 05:39:59.892087       1 authentication.go:220] Unable to get configmap/extension-apiserver-authentication in kube-system.  Usually fixed by 'kubectl create rolebinding -n kube-system ROLE_NAME --role=extension-apiserver-authentication-reader --serviceaccount=YOUR_NS:YOUR_SA'
F1127 05:39:59.892129       1 heapster.go:97] Could not create the API server: configmaps "extension-apiserver-authentication" is forbidden: User "system:serviceaccount:openshift-monitoring:metrics-server" cannot get configmaps in the namespace "kube-system": no RBAC policy matched
```

jim-minter replied:
thanks for checking - although it's strange that this error happens at all - do the sync pod logs show that the pod is being created, e.g. before the RBAC is?  If so that's a minor error.  Could you investigate - perhaps the filter configuration in `func (s *sync) writeDB()` needs to be adjusted to guarantee RBAC earlier in the sync run?

olga-mir replied:
yes, looked at the logs and Deployment is being created just before RoleBinding. I can see deployment and RoleBinding are in the "other resources" so that's why it is not guaranteed, could be an alphabetic order? https://github.com/openshift/openshift-azure/blob/0c718f237b8ea05224abd984bb6e4c26b15d345b/pkg/sync/v14/sync.go#L328-L392

olga-mir replied:
@jim-minter should we include this in this PR then https://github.com/openshift/openshift-azure/pull/2099/commits/ac8afce4f38f31861a80bd9e88e450ce14c4a876?
(I'll squash new commits before making this PR ready to review)

olga-mir commented:
these are the code changes + e2e test for this PR: https://github.com/openshift/openshift-azure/pull/2099/commits/d64734d46b99758fa980c7a6a5ef1700ac011feb

There is a merge conflict while rebasing to latest master because it also has vendor changes. I might open a new branch based on latest master and do vendor updates on top of it. But so far this approach didn't work, while running the same test (`FOCUS="\[MetricsServer\]" ./hack/e2e.sh`) I get this error: `pkg/api/validate/serviceprincipal.go:45:15: p.ParseUnverified undefined (type *jwt.Parser has no field or method ParseUnverified)`

olga-mir commented:
this is now rebased and ready for review. 
Tested `k top nodes` `k top po` works as expected.
Verified certificate served by metrics server:
```
...             
        Issuer: CN=openshift-signer                              
        Validity                                                 
            Not Before: Dec  1 23:40:12 2019 GMT                 
            Not After : Dec  1 23:40:12 2021 GMT                 
        Subject: CN=metrics-server.openshift-monitoring.svc      
....
```
Also added e2e tests pass

olga-mir commented on pkg/sync/v14/sync.go:
```diff
@@ -280,6 +280,12 @@ var (
+	}
+	roleBindingFilter = func(o unstructured.Unstructured) bool {
+		return o.GroupVersionKind().Kind == "ClusterRoleBinding" || o.GroupVersionKind().Kind == "RoleBinding"
+	}
```
without setting the order, metrics-server deployment is created before the role bindings and it will be in CrashLoopBackoff for a short time.  But the deployment is functional shortly after that, so these changes are not crucial

ehashman commented:
I was able to verify that, on a master node with the root kubeconfig, metrics server is running and I can pull metrics telemetry :tada:

```
[root@master-000000 ~]# oc get pod -n openshift-monitoring
NAME                                          READY     STATUS    RESTARTS   AGE
...
metrics-server-85d54ffcf8-pqkf6               1/1       Running   0          31m
...
[root@master-000000 ~]# kubectl top nodes
NAME                        CPU(cores)   CPU%      MEMORY(bytes)   MEMORY%   
compute-1575418179-000000   106m         6%        811Mi           12%       
infra-1575418179-000000     200m         12%       1998Mi          29%       
infra-1575418179-000001     306m         19%       1964Mi          29%       
infra-1575418179-000002     281m         17%       1526Mi          22%       
master-000000               300m         20%       1986Mi          29%       
master-000001               385m         25%       2179Mi          32%       
master-000002               410m         27%       2414Mi          36%       
[root@master-000000 ~]# kubectl top pods
NAME                                CPU(cores)   MEMORY(bytes)   
docker-registry-b9njv               1m           10Mi            
docker-registry-lnjsj               1m           11Mi            
docker-registry-zxzf6               1m           12Mi            
registry-console-7f6bb49958-djldc   2m           1Mi             
registry-console-7f6bb49958-v8kpf   2m           1Mi             
router-7n8qg                        5m           46Mi            
router-dc7zs                        9m           37Mi            
router-vnhcq                        6m           35Mi  
```

However, I haven't been able to get `kubectl top` to work as either a regular user or customer admin. @olga-mir how did you test that use case? As this is currently implemented, I have the suspicion that customers can't actually use the feature because they lack full cluster admin access.

```
ehashman@red-dot:~/go/src/github.com/openshift/openshift-azure$ oc whoami
ehashman@rhcuppettgmail.onmicrosoft.com
ehashman@red-dot:~/go/src/github.com/openshift/openshift-azure$ oc get groups
NAME                  USERS
osa-customer-admins   ... ehashman@rhcuppettgmail.onmicrosoft.com, ...
ehashman@red-dot:~/go/src/github.com/openshift/openshift-azure$ kubectl top nodes
Error from server (Forbidden): nodes.metrics.k8s.io is forbidden: User "ehashman@rhcuppettgmail.onmicrosoft.com" cannot list nodes.metrics.k8s.io at the cluster scope: no RBAC policy matched
ehashman@red-dot:~/go/src/github.com/openshift/openshift-azure$ kubectl top pods
Error from server (Forbidden): pods.metrics.k8s.io is forbidden: User "ehashman@rhcuppettgmail.onmicrosoft.com" cannot list pods.metrics.k8s.io in the namespace "openshift": no RBAC policy matched
```

ehashman commented on test/e2e/specs/metrics_server.go:
```diff
@@ -0,0 +1,37 @@
+		Expect(err).ToNot(HaveOccurred())
+		Expect(res.Containers[0].Usage.Cpu().MilliValue() > 0)
+		Expect(res.Containers[0].Usage.Memory().Value() > 0)
+	})
```
The e2e tests here appear to check that one can pull metrics with an admin config from the cluster, but I don't think that quite reflects the end-user behaviour we want to test; we want a customer admin to be able to deploy HPA. I'm fine with that test change coming in a follow-up PR, but I don't think this quite tests the use case we're building for yet.

olga-mir replied:
yes that's right. Customer admin do not have permission to list metrics or to deploy hpa as far as I can see, so I will have to add following rbac permissions to customer-admin-cluster (and -project?) and also change these tests. 
```
- apiGroups:
    - metrics.k8s.io
  resources:
    - pods
    - nodes
  verbs:
    - get
    - list
    - watch
```
and
```
- apiGroups:
  - autoscaling
  resources:
  - horizontalpodautoscalers
  verbs:
  - create
  - delete
  - deletecollection
  - get
  - list
  - patch
  - update
  - watch
```

ehashman replied:
Without those RBAC changes, customers won't actually be able to use the metrics server, so let's proceed with adding them.

olga-mir replied:
@ehashman I've discovered that allowing `kubectl top pods` to customer admin would allow them to see pods that otherwise they cannot see. Maybe we can block it per namespace?
For now this is commit which will allow to query nodes usage, but not pods and deploy HPA: https://github.com/openshift/openshift-azure/pull/2099/commits/d9f7155de03e1d2d7d1d5901e77a1271ae215331

I also want to test now HPA, which I haven't done yet.

ehashman replied:
> would allow them to see pods that otherwise they cannot see

Which pods? Stuff in kube-system or something?

I think it'd be ideal to figure out HPA now, since the reason we're adding metrics-server is to enable the HPA use case.

olga-mir replied:
I've tested the HPA (manually) - customer admin is able to deploy HPA and HPA scaling works too. As discussed offline, I will add a e2e test for deploying HPA as customer admin. 

For pods: Any namespace - as example bellow shows, the pod list is not available with `k get pods` but is available with `k top pods` (this is before I removed `pods` from the cluster role definition)

As discussed offline, we decided to enable to view `pods`, so I will add it back and update the tests accordingly.

```sh
[ openshift-azure (metrics-server-1) ] oc whoami                                                     
omirensk@rhcuppettgmail.onmicrosoft.com                                                                    
[ openshift-azure (metrics-server-1) ] k top pods -n openshift-azure-logging                            
NAME                                           CPU(cores)   MEMORY(bytes)
log-analytics-cluster-agent-5765c8dffb-p64fw   12m          95Mi  
log-analytics-node-agent-mvqn8                 6m           71Mi          
[ openshift-azure (metrics-server-1) ] k get pods -n openshift-azure-monitoring
No resources found.                                                                                                                                                                                             
Error from server (Forbidden): pods is forbidden: User "omirensk@rhcuppettgmail.onmicrosoft.com" cannot list pods in the namespace "openshift-azure-monitoring": no RBAC policy matched
```

olga-mir replied:
@ehashman 
the hpa section in the ClusterRole is not needed, customer admins are able to deploy HPAs without it in the non-openshift projects. Even in e2e test with `EndUser` the hpa seems to be deployed without any issue. So I didn't add another e2e for this bit. 

You can view the change in this commit: https://github.com/openshift/openshift-azure/pull/2099/commits/c71b95598b8bca2b61fa0fb0c5110531d7bf6d51

the following output demonstrates this for customer admin use-case:
```
$ oc whoami 
omirensk@rhcuppettgmail.onmicrosoft.com
$ k top pods
$ k top pods -n openshift-azure-logging
NAME                                           CPU(cores)   MEMORY(bytes)   
log-analytics-cluster-agent-67467699f5-989h9   16m          84Mi            
log-analytics-node-agent-mvqn8                 6m           79Mi            
$ 
$ k get hpa
NAME          REFERENCE                      TARGETS         MINPODS   MAXPODS   REPLICAS   AGE
welcome-php   DeploymentConfig/welcome-php   <unknown>/80%   1         5         0          4m
$ k delete hpa welcome-php
horizontalpodautoscaler.autoscaling "welcome-php" deleted
$
$ k get hpa
No resources found.
$
$ k apply -f tmp/welcome-php-hpa.yaml 
horizontalpodautoscaler.autoscaling/welcome-php created
$
$ k get hpa
NAME          REFERENCE                      TARGETS         MINPODS   MAXPODS   REPLICAS   AGE
welcome-php   DeploymentConfig/welcome-php   <unknown>/80%   1         5         0          4s
$
$ oc describe ClusterRole customer-admin-cluster
Name:                                           customer-admin-cluster
Created:                                        28 hours ago
Labels:                                         azure.openshift.io/owned-by-sync-pod=true
Annotations:                                    <none>
Verbs                                           Non-Resource URLs       Resource Names  API Groups                      Resources
[create delete get list patch update watch]     []                      []              [servicecatalog.k8s.io]         [clusterservicebrokers]
[create delete get list patch update watch]     []                      []              [ user.openshift.io]            [groups identities useridentitymappings users]
[create delete get list patch update watch]     []                      []              [ authorization.openshift.io]   [clusterrolebindings rolebindings]
[create delete get list patch update watch]     []                      []              [rbac.authorization.k8s.io]     [clusterrolebindings rolebindings]
[delete get list watch]                         []                      []              [ oauth.openshift.io]           [oauthclientauthorizations]
[create]                                        []                      []              [ authorization.openshift.io]   [resourceaccessreviews subjectaccessreviews]
[create]                                        []                      []              [authorization.k8s.io]          [subjectaccessreviews]
[create]                                        []                      []              [ project.openshift.io]         [projectrequests]
[get list watch]                                []                      []              []                              [events nodes persistentvolumes pod]
[create delete get list patch update watch]     []                      []              [ security.openshift.io]        [securitycontextconstraints]
[get list watch]                                []                      []              [ quota.openshift.io]           [clusterresourcequotas]
[get list]                                      []                      []              [ authorization.openshift.io]   [clusterpolicybindings]
[get list watch]                                []                      []              [ image.openshift.io]           [images imagestreamtags]
[get list update]                               []                      []              [ network.openshift.io]         [netnamespaces]
[get list]                                      []                      []              [ network.openshift.io]         [clusternetworks]
[get list watch]                                []                      []              [ build.openshift.io]           [buildconfigs builds]
[get list watch]                                []                      []              [metrics.k8s.io]                [nodes pods]
```

olga-mir commented:
some tests failed with `fatal: unable to access 'https://github.com/openshift/openshift-azure.git/': Could not resolve host: github.com` 

/retest

ehashman commented on pkg/sync/v14/data/ClusterRole.rbac.authorization.k8s.io/customer-admin-cluster.yaml:
```diff
@@ -171,20 +171,8 @@ rules:
-  - list
-  - patch
-  - update
-  - watch
```
@Makdaam can you PTAL?

olga-mir replied:
note, that this section I added earlier in this PR, it is not in diff against master: https://github.com/openshift/openshift-azure/pull/2099/files#diff-e024d6fd3fbeb1611a5f7f4178de9ae9

ehashman commented on test/e2e/specs/metrics_server.go:
```diff
@@ -29,13 +29,7 @@ var _ = Describe("Openshift on Azure metrics server tests [MetricsServer][EveryP
-		Expect(err.Error()).To(Equal("pods.metrics.k8s.io \"master-api-master-000000\" is forbidden: User \"customer-cluster-admin\" cannot get pods.metrics.k8s.io in the namespace \"kube-system\": no RBAC policy matched"))
-	})
-
-	It("should should return valid metrics a given pod when accessed as a cluster admin [MetricsServer]", func() {
```
I think this is the right test name and this needs to be updated?

olga-mir replied:
yes, the name is definitely wrong, thanks for catching it

ehashman commented:
Great work @olga-mir, this looks very close :100:

olga-mir commented:
/flake https://github.com/openshift/openshift-azure/issues/2126

/retest

ehashman commented:
Looks like there's a bindata conflict (probably due to content update), can you rebase?
